src/wasm-interpreter.h

```diff
@@ -2228,6 +2228,114 @@ class ModuleInstance
     : ModuleInstanceBase(wasm, externalInterface) {}
 };
 
+static const Name NONSTANDALONE_FLOW("Binaryen|nonstandalone");
+
+typedef std::unordered_map<LocalGet*, Literals> GetValues;
+
+// Evaluates a standalone expression. Errors if we hit anything that can't be
+// evaluated.
+class StandaloneExpressionRunner
+  : public ExpressionRunner<StandaloneExpressionRunner> {
+  Module* module;
+
+  // map gets to constant values, if they are known to be constant
+  GetValues& getValues;
+
+public:
+  // Whether we are trying to precompute down to an expression (which we can
+  // do on say 5 + 6) or to a value (which we can't do on a local.tee that
+  // flows a 7 through it). When we want to replace the expression, we can
+  // only do so when it has no side effects. When we don't care about
+  // replacing the expression, we just want to know if it will contain a known
+  // constant.
+  enum Intent {
+    // Intent is to evaluate the expression, so we can ignore some side effects
+    EVALUATE,
+    // Intent is to replace the expression, so side effects must be retained
+    REPLACE_EXPRESSION
+  };
+  Intent intent;
+
+  StandaloneExpressionRunner(Module* module,
+                             GetValues& getValues,
+                             Intent intent,
+                             Index maxDepth)
+    : ExpressionRunner<StandaloneExpressionRunner>(maxDepth), module(module),
+      getValues(getValues), intent(intent) {}
+
+  virtual ~StandaloneExpressionRunner() {}
```

Comment:
I don't know what I'm doing here btw, but this was necessary to silence an error when `delete`ing the runner again. That correct?

Reply:
Hmm, this is a little fishy because there is no inheritance or any other virtual methods here. What is the error you get if you delete this line?

Reply:
The error is

```
binaryen-c.cpp:4864:10: error: deleting object of polymorphic class type
‘wasm::StandaloneExpressionRunner’ which has non-virtual destructor might
cause undefined behavior [-Werror=delete-non-virtual-dtor]

   delete R;
          ^
```

Reply:
Ok, I figured out what was going on here. One of the parent classes of `ContextAwareExpressionRunner` has virtual methods, so the compiler pessimistically assumes that this class might have subclasses, so it requires a virtual destructor implementation. But if you change the declaration `class ContextAwareExpressionRunner` to `class ContextAwareExpressionRunner final`, you can tell the compiler that there will be no more subclasses and it will be able to use the default destructor. What a language.

Reply:
Great, that worked! :)